### PR TITLE
ci: refine job matrix and enable cache for poetry

### PR DESCRIPTION
**Pull request by Umpire2018**
- Replace direct installation of Poetry with pipx to ensure isolated environment setups.
- Enable caching for Poetry dependencies using setup-python action to improve build efficiency.
- Refactor the job matrix specifications.

> Use jobs.<job_id>.strategy.matrix.include to expand existing matrix configurations or to add new configurations. 

https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstrategymatrixinclude

### CODE REVIEW

**Comment by rbren** [comment]
LGTM! CC @libowen2121 for the matrix change

File: .github/workflows/run-integration-tests.yml [hunk 1]
```diff
@@ -6,58 +6,38 @@ jobs:
...
-            sandbox: "ssh"
-          - name: PlannerAgent-py311-ssh
-            python-version: "3.11"
-            agent: "PlannerAgent"
-            embedding-model: "none"
-            sandbox: "ssh"
-          - name: MonologueAgent-py311-ssh
-            python-version: "3.11"
-            agent: "MonologueAgent"
+          - agent: "MonologueAgent"
```
Comment (li-boxuan) [inline .github/workflows/run-integration-tests.yml:15]: Shouldn't we remove this line (and L17, L19, L21)?

Comment (Umpire2018) [reply .github/workflows/run-integration-tests.yml:15]: I don't quite follow.

Comment (li-boxuan) [reply .github/workflows/run-integration-tests.yml:15]: sorry, my bad; this looks good

**Review by li-boxuan — APPROVED** [review]
Mind solving the conflicts?

**Comment by Umpire2018** [comment]
> Mind solving the conflicts?

Done. 

It seems that enabling caching for `Homebrew` did not bring any time savings, and it introduced new issues, therefore, the introduction of caching have been cancelled.

**Comment by li-boxuan** [comment]
Something seems to be wrong? Integration CI tests don't show up now.

**Comment by Umpire2018** [comment]
> Something seems to be wrong? Integration CI tests don't show up now.

My bad. :sweat_smile:

image

But it's weird that [on-linux (MonologueAgent, local, local)](https://github.com/OpenDevin/OpenDevin/actions/runs/8858290102/job/24326747659?pr=1381#logs) failed but others succeeded.

**Comment by li-boxuan** [comment]
> But it's weird that [on-linux (MonologueAgent, local, local)](https://github.com/OpenDevin/OpenDevin/actions/runs/8858290102/job/24326747659?pr=1381#logs) failed but others succeeded.

CI complains about python version, but we do have `python-version: ${{ matrix.python-version }}` defined, so I have no idea what's happening.

Retry shall work, I guess?

**Review by Borda — APPROVED** [review]

**Comment by Umpire2018** [comment]
@rbren Passed tests and ready for review.

File: .github/workflows/run-unit-tests.yml [hunk 2]
```diff
@@ -35,13 +43,21 @@ jobs:
...
+      - name: Set up Python
+        uses: actions/setup-python@v5
         with:
           python-version: ${{ matrix.python-version }}
-      - name: Install Poetry
-        run: curl -sSL https://install.python-poetry.org | python3 -
+          cache: 'poetry'
+
+      - name: Install Python dependencies using Poetry
+        run: poetry install --without evaluation
```
Comment (li-boxuan) [inline .github/workflows/run-unit-tests.yml:57]: this is part of `make build`

Comment (Umpire2018) [reply .github/workflows/run-unit-tests.yml:57]: I know which have been removed in previous commit. I think add `poetry install` here can speed up `actions` and looks like no other side effects so i added it here.

Comment (li-boxuan) [reply .github/workflows/run-unit-tests.yml:62]: Logically it wouldn't make sense to speed up... but it wouldn't harm either. I feel like we could remove `make build` completely from run-integration-tests.yml though.

**Review by li-boxuan — APPROVED** [review]

**Review by rbren — APPROVED** [review]

**Pull request merged**